gitlab/utils.py

```diff
@@ -161,3 +161,28 @@ def warn(
         stacklevel=stacklevel,
         source=source,
     )
+
+
+def _validate_attrs(
+    data: Dict[str, Any],
+    attributes: types.RequiredOptional,
+    excludes: Optional[list] = None,
+) -> None:
+    if excludes is None:
+        excludes = []
+
+    required = [k for k in attributes.required if k not in excludes]
+    missing = [attr for attr in required if attr not in data]
+
+    if missing:
+        raise AttributeError(f"Missing attributes: {', '.join(missing)}")
+
+    exclusives = [attr for attr in data if attr in attributes.exclusive]
```

Comment:
I think need to early exit if `attributes.exclusive` is empty. Otherwise it fails.

Reply:
good catch. like this?

```python
if not attributes.exclusive:
    return
```

Reply:
That would do it. Thanks.

Reply:
what attributes.required .. what if none are required? same problem? perhaps this is better?

```python
    if attributes.required:
        required = [k for k in attributes.required if k not in excludes]
        missing = [attr for attr in required if attr not in data]

        if missing:
            raise AttributeError(f"Missing attributes: {', '.join(missing)}")

    if attributes.exclusive:
        exclusives = [attr for attr in data if attr in attributes.exclusive]
        if len(exclusives) > 1:
            raise AttributeError(
                f"Provide only one of these attributes: {', '.join(exclusives)}"
            )
        if not exclusives:
            raise AttributeError(
                f"Must provide one of these attributes: {', '.join(attributes.exclusive)}"
            )
```

Reply:
That would work.

Reply:
done